### DISCUSSION TITLE
Support shared configs

### DISCUSSION THREAD

**Discussion by anbraten** [woodpecker-ci/woodpecker#2286] [Request a Feature]
### Clear and concise description of the problem

As a user with multiple similar Woodpecker projects, I would like to share some configs between them. This would allow me to adjust configs by only editing it at one place.

### Suggested solution

- "importing" pipeline config templates via urls
- some kind of callable pipeline would be nice: a pipeline that is working like a function which will be called by some other pipeline by passing some arguments

### Alternative

In the meantime it would be possible to use #915 to use my own config pre-processing service which somehow injects shared configs.

### Validations

- [X] Read the [Contributing Guidelines](https://github.com/woodpecker-ci/woodpecker/blob/master/CONTRIBUTING.md).
- [X] Read the [docs](https://woodpecker-ci.org/docs/intro).
- [X] Check that there isn't already an [issue](https://github.com/woodpecker-ci/woodpecker/issues) that request the same feature to avoid creating a duplicate.

**Comment by benyanke** [comment]
Adding additional info from the now-closed-as-duplicate #1179:

As someone who has a number of small repos that all require similar CI config, I find gitlab's CI includes very useful.

This allows you to reference yaml documents in another repo and then they are included, and YAML merged, if multiples. This allows more composable pipelines, with more flexibility than the Woodpecker plugins (which are excellent, for some workflows).

gitlab include docs
https://docs.gitlab.com/ee/ci/yaml/includes.html

To use my own gitlab repos as an example:

I have a repo at gitlab.com/benyanke/common-ci-configs containing a number of composable CI bits that I use in various projects. One builds and pushes a docker container, one builds a hugo site, one publishes a hugo site to Netlify (a hosting provider), one publishes a hugo site to another provider, one lints, tests, and builds Golang projects, etc. On any given project, instead of needing to redefine all the required logic for a docker or go project, I simply include the go or docker include.

**Reply by wrenix** [reply]
Now gitlab has also components:
https://docs.gitlab.com/ci/components/

This is even more powerfull with variables/inputs and templating.

PS: Now means 22.05.2023 ... and is stabil and enabled everwhere since 16.05.2024
PSS: personal: I like every easy fast implemented way with is possible with allow us to use from other git-repo. But if we are equal to gitlab-ci i am also happy.

**Reply by anbraten** [reply]
That's a pretty smart way of re-using configs.

**Comment by JFronny** [comment]
I have attempted to implement something like this using the External Configuration API, but it seems that woodpecker fails to find pipelines using custom syntax, which my service is supposed to convert (adding an `include` key, in this case). If that is fixed, implementing this in a separate service would become trivial.
Edit: I have implemented a PoC of this feature as an external service that processes comments in the yaml file [here](https://git.frohnmeyer-wds.de/Johannes/Woodpecker-Include)

**Comment by genofire** [comment]
For me, that is really importent ... i am really thinking of switching to gitlab-ci again for that.

**Comment by 6543** [comment]
> For me, that is really importent ... i am really thinking of switching to gitlab-ci again for that.

that's sad but whatever works best for you :/

**Comment by anbraten** [comment]
https://docs.drone.io/template/

**Comment by Sh4kE** [comment]
What is the status of this feature? It's highly wanted and no progress since August 2023 :(

**Comment by victorsmits** [comment]
Hello, I need this feature ! whats the status of it ?

**Comment by qwerty287** [comment]
Please don't comment like this.

This is an open source project managed in our spare time. Nobody stops you from working on this issue and proposing changes, were happy to assist you if necessary.

However, if there's no comment, we just have other priorities. Asking about the status of this won't change this.

**Reply by victorsmits** [reply]
Sorry I did not want to bother you ! I am looking for this feature since a long time and we are planning to use this project in my main company. 

Yes I know this is an open source project but I do not know java, so I can not work on this project and I do not have time to work on this. 

I understand that my message was a bit aggressiv and I am sorry about that.

**Reply by zc-devs** [reply]
> I do not know java

Maybe you know `go`? 🤣

**Reply by victorsmits** [reply]
nope never try it ahaha

**Comment by RaphMad** [comment]
I created an external config service that pretty much mimics what the template feature in Drone did: https://github.com/RaphMad/woodpecker_template_config_provider

Note that template data is optional, so it can also be used to share "plain" configs between multiple projects by including a single-line file.

Using it as of `2025.12.15` / Woodpecker `v3.12.0` in my homelab environment.
